ds-querier: handle execute errors better

asimpson opened the pull request:
This addresses two problems:

1. We had lots of "false positive" errors because we'd return `responder.Error` with the error message instead of `responder.Object` with the `qdr` which contains the error message. https://github.com/grafana/grafana/blob/48c5b55cdb50780dc53ec6b9adf14e65e9275442/pkg/registry/apis/query/query.go#L177-L180
2. In a case where we get an error after querying we weren't returning an error qdr, just the err, which resulted in broken parsing of the err. https://github.com/grafana/grafana/blob/48c5b55cdb50780dc53ec6b9adf14e65e9275442/pkg/registry/apis/query/query.go#L215-L218

asimpson committed:
ds-querier: handle execute errors better

sarahzinger reviewed (approved):
This looks good to me! As a heads up it will break https://ops.grafana-ops.net/d/beea292n0sjy8e/error-logs-for-ds-querier?orgId=1&from=now-6h&to=now&timezone=browser&var-cluster=ops-eu-south-0&var-Filters= but that seems ok to me! We can just change the dashboard

sarahzinger commented on pkg/registry/apis/query/query.go:
```diff
@@ -160,21 +160,17 @@ func (r *queryREST) Connect(connectCtx context.Context, name string, _ runtime.O
-					b.log.Error("hit unexpected k8s error while executing query", "err", err, "status", k8sErr.Status())
-				}
-				b.log.Debug("sending a known k8s error to the client", "err", err, "status", k8sErr.Status())
+			b.log.Error("execute error", "http code", query.GetResponseCode(rsp), "err", err)
```
if you don't have a rsp will it error out?

sarahzinger replied:
maybe something like if rsp exists use that status code otherwise it's 500

asimpson replied:
if `rsp` is nil, `GetResponseCode` fills in `400`. I'll tweak how `GetResponseCode` works. Good shout!

sarahzinger commented on pkg/registry/apis/query/query.go:
```diff
@@ -272,6 +268,10 @@ func (b *QueryAPIBuilder) handleQuerySingleDatasource(ctx context.Context, req d
 	}
 
+	if err != nil {
+		rsp = buildErrorResponse(err, req)
```
maybe we should add a debug log line?

asimpson replied:
Good idea!

asimpson committed:
fix: change how GetResponseCode works to return 418 if rsp is nil

418 is a bit of an easter egg which in this case works since we don't
have an rsp but we do know something went wrong, so a 200 won't work.

Also changed this to return the code in the frame, not sure why we
weren't.

sarahzinger commented on pkg/apis/query/v0alpha1/query.go:
```diff
@@ -30,11 +30,11 @@ type QueryDataResponse struct {
 func GetResponseCode(rsp *backend.QueryDataResponse) int {
 	if rsp == nil {
-		return http.StatusBadRequest
+		return http.StatusTeapot // rsp is nil, so we return a teapot
```
While I love a teapot easter egg, will this mean we miss errors in our slo?

asimpson replied:
Our SLO is 500s over time so this won't change our SLO performance one way or another.

The teapot use is to indicate that these messages had no downstream status code. I realized we were previously masking real 400s we might get from the client or further downstream.

The new logic in https://github.com/grafana/grafana/pull/105496/commits/5574f4f4dbc3653c3f6fee0e996c0269bd2ede51 first checks for the combination of `Error AND Status`. If we have an `Error` but no `Status` we use a teapot.

sarahzinger replied:
I guess it will still get converted to a 500 by the responder.Error... but we'd be logging out teapots so we just have to know what that means haha. But I'm game

asimpson committed:
tests: fix GetResponseCode tests

sarahzinger reviewed (approved):
🚀

asimpson committed:
log no rsp case

asimpson committed:
bring back og error log
